drivers/i2s/i2s_nrfx.c

```diff
@@ -0,0 +1,1336 @@
+		}
+		ret = channel_drop(i2s, dir);
+		break;
+	}
+	default:
+		LOG_ERROR("Invalid trigger command",
+			  (u32_t)channel->current_state);
+		return -EINVAL;
+	}
+	if (ret < 0) {
+		LOG_ERROR("Error trigger while execution",
+			  (u32_t)channel->current_state);
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return ret;
+	}
+	channel->last_trigger_cmd = cmd;
+	return 0;
+}
+
+/*
+ * channel management functions
+ */
+
+static struct channel_str *channel_get(struct i2s_nrfx_data *i2s,
+						 enum i2s_dir dir)
+{
+	assert(i2s != NULL);
+	switch (dir) {
+	case I2S_DIR_RX:
+		return &i2s->channel_rx;
+	case I2S_DIR_TX:
+		return &i2s->channel_tx;
+	}
+	return NULL;
+}
+
+static int channel_change_state(struct channel_str *channel,
+				 enum i2s_state new_state)
+{
+	bool change_forbidden = false;
+	enum i2s_state old_state = channel->current_state;
+
+	assert(channel != NULL);
+	switch (new_state) {
+		break;
+	case I2S_STATE_READY:
+		if (old_state == I2S_STATE_READY) {
+			change_forbidden = true;
+		}
+		break;
+	case I2S_STATE_RUNNING:
+		if (old_state != I2S_STATE_READY) {
+			change_forbidden = true;
+		}
+		break;
+	case I2S_STATE_STOPPING:
+		if (old_state != I2S_STATE_RUNNING) {
+			change_forbidden = true;
+		}
+		break;
+	case I2S_STATE_NOT_READY:
+	case I2S_STATE_ERROR:
+		break;
+	default:
+		LOG_ERR("Invalid channel state chosen");
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return -EINVAL;
+	}
+
+	if (change_forbidden) {
+		LOG_ERROR("Failed to change channel state",
+			  (u32_t)channel->current_state);
+		channel_change_state(channel, I2S_STATE_ERROR);
+		return -EIO;
+	}
+	channel->current_state = new_state;
+	return 0;
+}
+
+static int channel_start(struct i2s_nrfx_data *i2s, enum i2s_dir dir)
+{
+	int ret;
+	unsigned int key;
+	struct channel_str * const channel_to_start = channel_get(i2s, dir);
+
+	ret = channel_change_state(channel_to_start, I2S_STATE_RUNNING);
+	if (ret < 0) {
+		return ret;
+	}
+	if (interface_get_state(i2s) != I2S_IF_RUNNING &&
+	    interface_get_state(i2s) != I2S_IF_READY) {
+		LOG_ERR("Invalid interface state");
+		return -EIO;
+	}
+	if (dir == I2S_DIR_RX) {
+		ret = k_mem_slab_alloc(channel_to_start->mem_slab,
+			(void **)&i2s->buffers.p_rx_buffer, K_NO_WAIT);
+	} else {
+		ret = channel_tx_get_data(i2s,
+			(u32_t **)&i2s->buffers.p_tx_buffer);
+	}
+	if (ret < 0) {
+		LOG_ERROR(dir == I2S_DIR_RX ? "Memory allocation error" :
+				      "Queue fetching error",
+				      (u32_t)channel_to_start->current_state);
+		return ret;
+	}
+	key = irq_lock();
+	if (interface_get_state(i2s) == I2S_IF_RUNNING) {
+		ret = interface_restart(i2s);
+	} else if (interface_get_state(i2s) == I2S_IF_READY) {
+		ret = interface_start(i2s);
+	}
+	irq_unlock(key);
+	if (ret < 0) {
+		LOG_ERROR("Failed to start/restart interface",
+			  (u32_t)channel_to_start->current_state);
+		return ret;
+	}
+	return 0;
+}
+
+static void channel_tx_mem_clear(struct i2s_nrfx_data *i2s,
+				 void const *first_block)
```

Comment:
This function is called only from two places, and only once it is called with `first_block` not being `NULL`. Why not free this first block right there, get rid of this parameter, and thus seriously simplify whole this function?